File: projectEuler010.py
# Summation of primes
#
# The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.
#
# Find the sum of all the primes below two million.

# Initialize a list, primes, to hold all the primes.

primes = []

def lowest(i):
    """Find the lowest whole number the input number is divisible by."""
    n = 2
    while i % n != 0:
        n = n+1
    return n

def isPrime(i):
    """Determine if the input number is prime."""
    if i == lowest(i):
        return True
    else:
        return False

# Step through the numbers from 2 to 2000000 and see if they're prime.
# If so, add them to the primes[] list.

for i in xrange(2,2000000):
    if isPrime(i) == True:
        primes.append(i)

# Add up everything in the primes[] list and print out the sum.

print sum(primes)

# I'm not at all happy with this. I know that this code works, per se,
# as I tried it with the example numbers given, finding all the primes
# below 10 and adding them up. Which it did with much aplomb. However,
# as the range increases, so does the time it takes to evaluate it all. 
# I let this run for half an hour and it still didn't complete. So
# obviously, there's some sort of trick to this that I'm not seeing, or
# perhaps an well-known algorithm that I'm not familiar with. Because,
# while the code here technically works, it's obviously the wrong solution.
# I will continue to explore this problem in more depth and revise as my
# understanding of the maths involved here grows. 

File: c/projectEuler010.c
/* Find the sum of all the primes below two million. */

#include <stdio.h>
#include <stdbool.h>
#include <math.h>

bool is_prime(unsigned long);
unsigned long sum_of_primes(long);

int main(void)
{
    printf("%d\n", sum_of_primes(2000000));
    return 0;
}

/* Determine if the input number is prime  */
bool is_prime(unsigned long input_num)
{
    if (input_num == 1)
    {
        return false;
    }
    else if (input_num == 2 || input_num == 3 || input_num == 5 || input_num == 7)
    {
        return true;
    }
    else if (input_num % 2 == 0)
    {
        return false;
    }
    else if (input_num % 3 == 0)
    {
        return false;
    }
    else if (input_num % 5 == 0)
    {
        return false;
    }
    else if (input_num % 7 == 0)
    {
        return false;
    }
    else
    { 
        return true;
    }
}

/* Calculate sum of primes below a given number */
unsigned long sum_of_primes(long max)
{
    unsigned long i;
    unsigned long sum = 0;

    for (i = 2; i < max; i++)
    {
        if (is_prime(i))
        {
            sum += i;
        }
    }

    return sum; 

}
